From bd718f249d1c74464a4f408ac7abfeca2dd32af5 Mon Sep 17 00:00:00 2001
From: Masayuki Nakagawa <nakagawa.msy@ncos.nec.co.jp>
Date: Wed, 19 May 2010 03:06:45 -0400
Subject: [PATCH] TCP: skb is unexpectedly freed Backport by lly from
 wl500g.googlecode.com

kernel.org commit fb7e2399ec17f1004c0e0ccfd17439f8759ede01
---
 net/ipv4/tcp_input.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/net/ipv4/tcp_input.c b/net/ipv4/tcp_input.c
index 5b3ee44f8d..fa97086d84 100644
--- a/net/ipv4/tcp_input.c
+++ b/net/ipv4/tcp_input.c
@@ -4600,9 +4600,11 @@ int tcp_rcv_state_process(struct sock *sk, struct sk_buff *skb,
 			 * But, this leaves one open to an easy denial of
 		 	 * service attack, and SYN cookies can't defend
 			 * against this problem. So, we drop the data
-			 * in the interest of security over speed.
+			 * in the interest of security over speed unless
+			 * it's still in use.
 			 */
-			goto discard;
+			kfree_skb(skb);
+			return 0;
 		}
 		goto discard;
 
-- 
2.11.4.GIT

